README.md
PHP Nano Rest
=============

[![Latest Stable Version](https://poser.pugx.org/gino-pane/nano-rest/v/stable)](https://packagist.org/packages/gino-pane/nano-rest)
[![Build Status](https://travis-ci.org/GinoPane/php-nano-rest.svg?branch=master)](https://travis-ci.org/GinoPane/php-nano-rest)
[![Maintainability](https://api.codeclimate.com/v1/badges/f87cf0eef8aad99c488c/maintainability)](https://codeclimate.com/github/GinoPane/php-nano-rest/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/f87cf0eef8aad99c488c/test_coverage)](https://codeclimate.com/github/GinoPane/php-nano-rest/test_coverage)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/GinoPane/php-nano-rest/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/GinoPane/php-nano-rest/?branch=master)
[![SensioLabs Insight](https://img.shields.io/sensiolabs/i/a31125f1-ff97-41c9-b0f1-9e6b5eb58470.svg)](https://insight.sensiolabs.com/projects/a31125f1-ff97-41c9-b0f1-9e6b5eb58470)
[![License](https://poser.pugx.org/gino-pane/nano-rest/license)](https://packagist.org/packages/gino-pane/nano-rest)
[![Total Downloads](https://poser.pugx.org/gino-pane/nano-rest/downloads)](https://packagist.org/packages/gino-pane/nano-rest)

Easy-to-use self-containing lightweight package to deal with cURL requests.

Such packages as [Guzzle](https://github.com/guzzle/guzzle) are great of course, but they are too heavy for small projects. You don't need an overkill package for some easy stuff, and that's when something small and neat may help.

Requirements
============

* PHP >= 7.1;
* cURL extension.

Installation
============

Require the package using command line:

    composer require "gino-pane/nano-rest:1.*"

or just put a new dependency in your existing `composer.json` and run `composer install` after that:


    "require": {
        ...
        "gino-pane/nano-rest": "1.*"
    }


Usage
=====

Project's philosophy implies usage of `RequestContext` and `ResponseContext` objects. `RequestContext` aggregates request settings
whilst `ResponseContext` contains response data.

Response context can be typed. Currently only `JsonResponseContext` available for JSON responses. Response type must be set explicitly by
user. If no response type set, `DummyResponseContext` will be used.

Please take a look at examples below, which may clarify everything.

#### POST some data to endpoint


    require './vendor/autoload.php';
    
    $nanoRest = new NanoRest();
    
    //create request context
    $requestContext = (new RequestContext('http://httpbin.org/post')) //pass URL to constructor
        ->setMethod(RequestContext::METHOD_POST) //set request method. GET is default
        ->setRequestParameters([ //set some request parameters. They will be attached to URL
            'foo' => 'bar'
        ])
        ->setData('Hello world!') //set request data for body
        ->setContentType(RequestContext::CONTENT_TYPE_TEXT_PLAIN) //being set by default
        ->setHeaders([ // set some headers for request
            'bar' => 'baz'
        ])
        ->setResponseContextClass(JsonResponseContext::class); //explicitly set expected response type
    
    $responseContext = $nanoRest->sendRequest($requestContext);
    
    $responseContext->getHttpStatusCode(); //200
    $responseContext->hasHttpError() //false
    
    $responseContext->getArray();
    
    /**
    array(8) {
      'args' =>
      array(1) {
        'foo' =>
        string(3) "bar"
      }
      'data' =>
      string(12) "Hello world!"
      'files' =>
      array(0) {
      }
      'form' =>
      array(0) {
      }
      'headers' =>
      array(8) {
        'Accept' =>
        string(3) "*/*"
        'Accept-Encoding' =>
        string(13) "deflate, gzip"
        'Bar' =>
        string(3) "baz"
        'Connection' =>
        string(5) "close"
        'Content-Length' =>
        string(2) "12"
        'Content-Type' =>
        string(25) "text/plain; charset=UTF-8"
        'Host' =>
        string(11) "httpbin.org"
        'User-Agent' =>
        string(13) "php-nano-rest"
      }
      'json' =>
      NULL
      'origin' =>
      string(12) "93.85.47.181"
      'url' =>
      string(31) "http://httpbin.org/post?foo=bar"
    }
    */


`RequestContext` provides `setCurlOption`/`setCurlOptions` which allow to override default cURL options
and customize request for all your needs. Please examine source code and provided `IntegrationTest` carefully
to get the whole idea.

#### Change the way how request query is generated

By default `http_build_query` encodes arrays using PHP square brackets syntax, like this:

    ?text[0]=1&text[1]=2&text[2]=3

But sometimes you'll want it to work like this instead:

    ?text=1&text=2&text=3

Or even in some other custom-defined way.

That's why `setEncodeArraysUsingDuplication` and `setHttpQueryCustomProcessor` methods were added to `RequestContext`:


    $url = "http://some.url";
    $data = ['text' => [1,2,3]];
    
    $request = (new RequestContext($url))
                ->setRequestParameters($data)
                ->setEncodeArraysUsingDuplication(false);
    
    $requestUrl = $request->getRequestUrl(); //http://some.url?text%5B0%5D=1&text%5B1%5D=2&text%5B2%5D=3
    
    $request = (new RequestContext($url))
                ->setRequestParameters($data)
                ->setEncodeArraysUsingDuplication(true);
    
    $requestUrl = $request->getRequestUrl(); //http://some.url?text=1&text=2&text=3
    

Method `setHttpQueryCustomProcessor` allows you to set your custom `Closure` that will be called on HTTP query string so you could process it as you wish. Initial request `$data` array will be passed to it as a second parameter.

    $url = "http://some.url";
    $data = ['text' => [1,2,3]];
    
    $request = (new RequestContext($url))
                ->setRequestParameters($data)
                ->setEncodeArraysUsingDuplication(true);
                
    $request->setHttpQueryCustomProcessor(
        function (string $query, array $data) {
            return str_replace('text', 'data', $query);
        }
    );
    
    $requestUrl = $request->getRequestUrl(); //http://some.url?data=1&data=2&data=3


Useful Tools
============

Running Tests:
--------

    php vendor/bin/phpunit
 
 or 
 
    composer test

Code Sniffer Tool:
------------------

    php vendor/bin/phpcs --standard=PSR2 src/
 
 or
 
    composer psr2check

Code Auto-fixer:
----------------

    php vendor/bin/phpcbf --standard=PSR2 src/ 
    
 or
 
    composer psr2autofix
 
 
Building Docs:
--------

    php vendor/bin/phpdoc -d "src" -t "docs"
 
 or 
 
    composer docs
    
Updating Cacert.pem:
--------
    
        php bin/update-cacert.php
     
     or
     
        composer update-cacert


Changelog
=========

To keep track, please refer to [CHANGELOG.md](https://github.com/GinoPane/php-nano-rest/blob/master/CHANGELOG.md).

Contributing
============

Please refer to [CONTRIBUTING.md](https://github.com/GinoPane/php-nano-rest/blob/master/CONTRIBUTING.md).

License
=======

Please refer to [LICENSE](https://github.com/GinoPane/php-nano-rest/blob/master/LICENSE).

Notes
=====

Powered by [composer-package-template](https://github.com/GinoPane/composer-package-template)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

Types of changes

* **Features** for initial package features.
* **Added** for new features.
* **Changed** for changes in existing functionality.
* **Deprecated** for soon-to-be removed features.
* **Removed** for now removed features.
* **Fixed** for any bug fixes.
* **Security** in case of vulnerabilities.

## [Unreleased]

## 1.2.3 - 2018-02-13

### Changed
* added composer script to update cacert.pem automatically after composer install or update.

## 1.2.2 - 2018-02-05

### Changed
* updated cacert.pem.

## 1.2.1 - 2018-01-11

### Changed
* Allow empty JSON response bodies to be processed correctly;
* renamed ResponseContext to ResponseContextAbstract (because it is abstract);
* renamed CONTRIBUTION.md to CONTRIBUTING.md;
* small README update.

### Added
* CODE_OF_CONDUCT.md, PULL_REQUEST_TEMPLATE.md, issue_template.md

## 1.2.0 - 2017-12-18

### Changed
* Response context handling was removed from NanoRest;
* response context handling was added to RequestContext instead.

## 1.1.3 - 2017-12-14

### Changed
* ROOT_DIRECTORY is defined in NanoRest namespace to avoid conflicts.

## 1.1.2 - 2017-12-12

### Added
* ResponseContext::getHttpStatusMessage which returns HTTP status message.

## 1.1.1 - 2017-12-12

### Added
* Started using NanoHttpStatus to detect HTTP error status;
* composer scripts.

### Changed
* Updated README.

## 1.1.0 - 2017-12-05

### Added
* RequestContext::setEncodeArraysUsingDuplication to build proper query strings in some cases.
* RequestContext::setHttpQueryCustomProcessor to add custom post-processor for query string.

### Changed
* All usages of "uri" were replaced by "url" for consistency.

## 1.0.1 - 2017-12-04

### Added

* Applied Scrutinizer badge.

### Fixed

* Fixed some typos;
* refactored some code for Scrutinizer.

### Security

* Removed error suppression for json_decode in JsonResponseContext

## 1.0.0 - 2017-11-30

### Features

* The initial release of PHP Nano Rest;
* almost all types of synchronous CURL requests are available;
* SSL is being verified using ca-cert bundle which is also included;
* code is fully covered with tests;
* generation of docs failed - PhpDoc cannot build full class hierarchy;
* more to come - short syntax for most common requests, cookies, response type based on content-type header, etc.

[Unreleased]: https://github.com/GinoPane/php-nano-rest/compare/v1.2.3...HEAD


src/Supplemental/CurlHelper.php
<?php

namespace GinoPane\NanoRest\Supplemental;

use GinoPane\NanoRest\{
    Request\RequestContext,
    Exceptions\TransportException
};

/**
 * Class CurlHelper
 */
class CurlHelper
{
    /**
     * Get a customized request handler to perform calls
     *
     * @param RequestContext $context
     *
     * @return resource
     */
    public function getRequestHandle(RequestContext $context)
    {
        $curlHandle = curl_init();

        $defaults = $this->getDefaultSettings($context);

        $defaults += $this->getCurlTimeoutSettings($context);

        $defaults += $this->getCurlSslSettings();

        $defaults += $this->getProxySettings($context);

        curl_setopt_array(
            $curlHandle, //@codeCoverageIgnore
            $context->getCurlOptions() + $defaults + $this->getRequestDataAndMethodOptions($context)
        );

        return $curlHandle;
    }

    /**
     * Execute curl handle
     *
     * @param resource $curlHandle
     *
     * @throws TransportException
     *
     * @return mixed
     */
    public function executeRequestHandle($curlHandle)
    {
        curl_setopt($curlHandle, CURLOPT_VERBOSE, true);
        $verbose = fopen('php://temp', 'w+');
        curl_setopt($curlHandle, CURLOPT_STDERR, $verbose);

        @list($headers, $response) = explode("\r\n\r\n", curl_exec($curlHandle), 2);

        $error = curl_error($curlHandle);
        $errorNumber = curl_errno($curlHandle);
        $httpStatus = curl_getinfo($curlHandle, CURLINFO_HTTP_CODE);

        curl_close($curlHandle);

        $this->handleCurlError($error, $errorNumber, $verbose);

        return [
            'response' => $response,
            'httpStatus' => $httpStatus,
            'headers' => $headers
        ];
    }

    /**
     * Get SSL settings for CURL handler
     *
     * CA certificate bundle was generated at Wed Jan 17 04:12:05 2018 GMT.
     * CURL_SSL_XXX options has recommended values for production environments
     *
     * @link https://curl.haxx.se/docs/caextract.html
     *
     * @link https://curl.haxx.se/libcurl/c/CURLOPT_SSL_VERIFYHOST.html
     * @link https://curl.haxx.se/libcurl/c/CURLOPT_SSL_VERIFYPEER.html
     *
     * @return array
     */
    private function getCurlSslSettings(): array
    {
        return [
            CURLOPT_SSL_VERIFYPEER  => true,
            CURLOPT_SSL_VERIFYHOST  => 2,
            CURLOPT_CAINFO          => \GinoPane\NanoRest\ROOT_DIRECTORY . DIRECTORY_SEPARATOR . 'cacert.pem'
        ];
    }

    /**
     * Get proxy settings for cURL handle
     *
     * @param RequestContext $context
     * @return array
     */
    private function getProxySettings(RequestContext $context): array
    {
        $proxy = [];

        if (!is_null($context->getProxy())) {
            $proxy[CURLOPT_PROXY] = $context->getProxy();
        }

        return $proxy;
    }

    /**
     * Get timeout settings for CURL handler
     *
     * @link https://curl.haxx.se/libcurl/c/CURLOPT_TIMEOUT_MS.html
     * @link https://curl.haxx.se/libcurl/c/CURLOPT_CONNECTTIMEOUT_MS.html
     *
     * @param RequestContext $context
     *
     * @return array
     */
    private function getCurlTimeoutSettings(RequestContext $context): array
    {
        $timeoutOptions = [];

        $timeout = $context->getTimeout();

        $timeoutOptions += $this->fillTimeoutOptions($timeout, CURLOPT_TIMEOUT, CURLOPT_TIMEOUT_MS);

        $connectionTimeout = $context->getConnectionTimeout();

        $timeoutOptions += $this->fillTimeoutOptions(
            $connectionTimeout, //@codeCoverageIgnore
            CURLOPT_CONNECTTIMEOUT,
            CURLOPT_CONNECTTIMEOUT_MS
        );

        return $timeoutOptions;
    }

    /**
     * Get defaults settings for cURL handle
     *
     * @param RequestContext $context
     * @return array
     */
    private function getDefaultSettings(RequestContext $context): array
    {
        $defaults = [
            CURLOPT_ENCODING => "",
            CURLOPT_USERAGENT => "php-nano-rest",
            CURLOPT_HEADER => true,
            CURLOPT_HTTPHEADER => array_values($context->getRequestHeaders()),
            CURLOPT_RETURNTRANSFER => true,
        ];

        return $defaults;
    }

    /**
     * @param RequestContext $context
     *
     * @return array
     */
    private function getRequestDataAndMethodOptions(RequestContext $context)
    {
        $curlOptions = array();

        $requestData = $context->getRequestData();

        $url = $context->getRequestUrl();

        switch ($context->getMethod()) {
            case RequestContext::METHOD_GET:
                $curlOptions[CURLOPT_HTTPGET] = 1;
                $url = $context->attachQueryToUrl($url, $requestData);
                break;
            case RequestContext::METHOD_POST:
                $curlOptions[CURLOPT_POST] = 1;
                $curlOptions[CURLOPT_POSTFIELDS] = $requestData;
                break;
            default:
                $curlOptions[CURLOPT_CUSTOMREQUEST] = $context->getMethod();
                $curlOptions[CURLOPT_POSTFIELDS] = $requestData;
        }

        $curlOptions[CURLOPT_URL] = $url;

        return $curlOptions;
    }

    /**
     * @param $error
     * @param $errorNumber
     * @param $verbose
     *
     * @throws TransportException
     */
    private function handleCurlError($error, $errorNumber, $verbose): void
    {
        if (!$error) {
            return;
        }

        $errorMessage = "\ncURL transport error: $errorNumber - $error";

        $transportException = new TransportException($errorMessage);

        rewind($verbose);

        $verboseLog = stream_get_contents($verbose);

        if ($verboseLog !== false) {
            $verboseLog = htmlspecialchars($verboseLog);

            $transportException->setData($verboseLog);
        }

        throw $transportException;
    }

    /**
     * Fill timeout options
     *
     * @param $timeout
     * @param $optionName
     * @param $optionNameMs
     *
     * @return array
     */
    private function fillTimeoutOptions($timeout, $optionName, $optionNameMs): array
    {
        $timeoutOptions = [];

        if (is_int($timeout)) {
            $timeoutOptions[$optionName] = $timeout;
        } elseif (is_float($timeout)) {
            $timeoutOptions[$optionNameMs] = $timeout * 1000;
            $timeoutOptions[CURLOPT_NOSIGNAL] = 1;
        }

        return $timeoutOptions;
    }
}
